sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c

```diff
@@ -764,61 +768,202 @@ static void set_codec_init_func(const struct snd_soc_acpi_link_adr *link,
+	 * to avoid generating CPU DAIs for this group again.
+	 */
+	group_generated[*group_id] = true;
+	*cpu_dai_num = index;
+
+	return 0;
+}
+
+/* convert 1 -9 to '1' - '9' and A -F to 'A' - 'F' */
+static inline int int_to_ascii(int x)
+{
+	return x > 9 ? x + 87 : x + 48;
+}
+
+static int create_sdw_dailink(struct device *dev,
+			      struct snd_soc_dai_link *dai_links,
+			      int *be_id, int sdw_be_num, int sdw_cpu_dai_num,
+			      struct snd_soc_dai_link_component *cpus,
+			      const struct snd_soc_acpi_link_adr *link,
+			      int *cpus_id, bool *group_generated)
 {
-	struct snd_soc_dai_link_component *codec;
-	unsigned int part_id, link_id;
-	int i, j = 0, idx;
+	const struct snd_soc_acpi_link_adr *link_next;
+	struct snd_soc_dai_link_component *codecs;
+	int cpu_dai_id[SDW_MAX_DEVICES];
+	int cpu_dai_num, cpu_dai_index;
+	unsigned int part_id, group_id;
+	int codec_idx = 0;
+	int i, j = 0, k, idx;
+	int codec_num;
 	int ret;
 
-	codec = devm_kcalloc(dev, link->num_adr,
-			     sizeof(struct snd_soc_dai_link_component),
-			     GFP_KERNEL);
-	if (!codec)
+	ret = get_slave_info(link, cpu_dai_id, &cpu_dai_num, &codec_num,
+			     &group_id, group_generated);
+	if (ret)
+		return ret;
+
+	codecs = devm_kcalloc(dev, codec_num, sizeof(*codecs), GFP_KERNEL);
+	if (!codecs)
 		return -ENOMEM;
 
-	ret = create_codec_dai_name(dev, link, codec);
-	if (ret < 0)
-		return ret;
+	link_next = link;
+
+	/* generate codec name on different links in the same group */
+	for (i = 0; link_next && link_next->num_adr && i < cpu_dai_num;
+		link_next++) {
+		const struct snd_soc_acpi_endpoint *endpoints;
+
+		endpoints = link_next->adr_d->endpoints;
+		if (group_id && (!endpoints->aggregated ||
+				 endpoints->group_id != group_id))
+			continue;
 
+		/* skip the link excluded by this processed group */
+		if (cpu_dai_id[i] != ffs(link_next->mask) - 1)
+			continue;
+
+		ret = create_codec_dai_name(dev, link_next, codecs, codec_idx);
+		if (ret < 0)
+			return ret;
+
+		/* check next link to create codec dai in the processed group */
+		i++;
+		codec_idx += link_next->num_adr;
+	}
+
+	/* find codec info to create BE DAI */
 	part_id = SDW_PART_ID(link->adr_d[0].adr);
 	idx = find_codec_info_part(part_id);
 	if (idx < 0)
 		return idx;
 
-	link_id = ffs(link->mask) - 1;
-
-	/* playback & capture */
-	for (i = 0; i < 2; i++) {
+	cpu_dai_index = *cpus_id;
+	for (i = SNDRV_PCM_STREAM_PLAYBACK;
+		i <= SNDRV_PCM_STREAM_CAPTURE; i++) {
+		char stream_id[SDW_MAX_DEVICES * 2];
 		char *name, *cpu_name;
 		static const char * const sdw_stream_name[] = {
-			"SDW%d-Playback",
-			"SDW%d-Capture",
+			"SDW%s-Playback",
+			"SDW%s-Capture",
 		};
 
 		if (!codec_info_list[idx].direction[i])
 			continue;
 
+		/* generate stream name with pattern SDW1:2:3 */
+		for (k = 0; k < cpu_dai_num; k += 1) {
+			stream_id[k * 2] = int_to_ascii(cpu_dai_id[k]);
+			stream_id[k * 2 + 1] = ':';
+
+			if (k == cpu_dai_num - 1)
+				stream_id[k * 2 + 1] = 0;
+		}
```

Comment:
do we actually need all this? Or can we name the stream

SDW1-Playback
SDW2-capture

where we have a linear set of dai links?

Please explain also the impact on topology.

Reply:
@bardliao , How do you implement your topology ?

Reply:
@plbossart That was my idea. I hope we can get all CPU ids for the stream name in multi cpu case. For example, "SDW1-Playback" for playback on sdw link 1 and "SDW1:2 Playback" for playback on sdw link 1 and 2.

Reply:
ah ok, that's a good idea, thanks @bardliao . All we need is a good comment on the 'pattern'